src/vm/i386/stublinkerx86.cpp

```diff
@@ -6428,16 +6428,36 @@ TADDR FixupPrecode::GetMethodDesc()
 #endif
 
 #ifdef FIXUP_PRECODE_PREALLOCATE_DYNAMIC_METHOD_JUMP_STUBS
+PCODE FixupPrecode::GetDynamicMethodPrecodeFixupJumpStub()
+{
+    WRAPPER_NO_CONTRACT;
+    _ASSERTE(((PTR_MethodDesc)GetMethodDesc())->IsLCGMethod());
+
+    // The precode fixup jump stub is shared by all fixup precodes in a chunk, and immediately follows the MethodDesc. Jump
+    // stubs cannot be reused currently:
+    //   - The jump stub's target would change separately from the precode being updated from "call Func" to "jmp Func", both
+    //     changes would have to be done atomically with runtime suspension, which is not done currently
+    //   - When changing the entry point from one version of jitted code to antoher, the jump stub's target pointer is not
+    //     aligned to 8 bytes in order to be able to do an interlocked update of the target address
+    // So, when initially the precode intends to be of the form "call PrecodeFixupThunk", if the target address happens to be
+    // too far for a relative 32-bit jump, it will use the shared precode fixup jump stub. When changing the entry point to
+    // jitted code, the jump stub associated with the precode is patched (for the first and last time), and the precode is
```

Comment:
I think the "first and last time" part is not exactly correct.

The jump stub can be rewritten once each time its associated DynamicMethodDesc gets reused (where reuse starts with the LinkMethod call in LCGMethodResolver::Destroy, and finishes when DynamicMethodTable::GetDynamicMethod dequeues the linked method and calls MethodDesc::Reset).

MethodDesc::Reset calls FixupPrecode::Init, which points the precode back to the GetDynamicMethodPrecodeFixupJumpStub() jump stub.  The next FixupPrecode::SetTargetInterlocked call (after the new LCG method is JIT'ed) will rewrite the GetDynamicMethodEntryJumpStub() jump stub, which could definitely hold the JIT'ed address of the previous LCG method at the time.

Reply:
Hmm let me think on this. As long as it's going to write the same target address to the jump stub we should be ok but if it may write a different address there could be issues.

Reply:
The single-write part is not true anyway, I tried to assert that and noticed that concurrent calls can call SetTargetInterlocked() multiple times but for the same target address.

Reply:
I think it's OK.  After MethodDesc::Reset, the jump stub isn't visible to anyone.  The upcoming calls to FixupPrecode::SetTargetInterlocked will all try to point the jump stub to the same address (i.e., the compiled body of the new LCG method).  No one will see the jump stub until it points to the new address, and no one will move the jump stub away from the new address until after the next MethodDesc::Reset (which doesn't happen until the MethodDesc is "dead", at which point the jump stub is once again not visible to anyone).

Reply:
Interesting, I didn't realize that those method descs get reused like this, always thought it was a dealloc-realloc coincidence. It looks ok to me as well. Also any call to Init or Reset is inherently unsafe if there can be any callers at the time. I'll update the relevant comments.